### PR TITLE
add CallSC support

### PR DESCRIPTION
**Pull request by damip**
# Problem

Being able to easily call a smart contract-exposed function without having to go through the pain of compiling a wasm ExecuteSC to do this simple and common task.

# Solution

Add support for the CallSC operation type that looks like this:

```rust
    /// Calls an exported function from a stored smart contract
    CallSC {
        /// Target smart contract address
        target_addr: Address,
        /// Target function name. No function is called if empty.
        target_func: String,
        /// Parameter to pass to the target function
        param: String,
        /// The maximum amount of gas that the execution of the contract is allowed to cost.
        max_gas: u64,
        /// Extra coins that are spent from the caller's sequential balance and transferred to the target
        sequential_coins: Amount,
        /// Extra coins that are spent from the caller's parallel balance and transferred to the target
        parallel_coins: Amount,
        /// The price per unit of gas that the caller is willing to pay for the execution.
        gas_price: Amount,
    },
```

And does the following:
* consensus substracts gas_price * max_gas + sequential_coins from the sequential balance of the sender, or else the op is invalid
* on execution, we start by crediting gas_price * max_gas to the block producer
* then we credit sequential_coins to the parallel balance of the sender (so that if everything else fails, the coins are not lost)
* then we enter the cancellable section (either everything in this list succeeds or nothing has a consequence):
  * init the call stack with only the caller address (to give it access rights to perform the following coin transfer)
  * transfer sequential_coins + parallel_coins from the sender's parallel balance to the target's parallel balance
  * add the target to the call stack
  * set context.coins = sequential_coins + parallel_coins
  * call the function target_func with the single string parameter param, under max_gas

### CODE REVIEW

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [inline massa-execution-worker/src/execution.rs:292]:
```suggestion
        // call the execution process specific to the operation type
```

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [inline massa-execution-worker/src/execution.rs:468]: Why are you crediting the operation sender? Shouldn't he be debited?

Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:468]: Ah it does give back on failure because of the context snapshot, ok nevermind.

**Comment by Eitu33** [comment]
I have taken extra time this once to make sure I don't miss anything but I'd rather have a second review tbh 😅

**Review by Eitu33 — APPROVED** [review]

**Comment by damip** [comment]
bors merge

**Pull request merged**
